WIP: Java BehaviorBuilder (#22748)

raboof opened the pull request:
Tests, docs and support for predicates still missing ;)

Partially based on https://github.com/akka/akka/compare/master...johanandren:play-around-with-typed-pre-2.5-rc1?expand=1

raboof committed:
First prototype (##22748)

Tests, docs and support for predicates still missing ;)

Partially based on https://github.com/akka/akka/compare/master...johanandren:play-around-with-typed-pre-2.5-rc1?expand=1

ktoso commented:
Refs #22748

ktoso commented on akka-typed-tests/src/test/java/akka/typed/javadsl/BehaviorBuilderTest.java:
```diff
@@ -0,0 +1,35 @@
+
+    @Test
+    public void shouldCompile() {
+        Behavior<Message> b = BehaviorBuilder.<Message>create()
```
I wonder if the create() is needed.
Perhaps we can just

```java
BehaviorBuilder
  .message(One.class, (ctx, o) -> ...
```

?

ktoso replied:
One could do so by mirroring the `message/signal` messages as static ones too, and those would be an entry point to the builder then.

patriknw replied:
or put the behaviorBuilder factory in Actor, similar i AbstractActor. Reduce number of imports.

ktoso replied:
good idea too!

ktoso commented on akka-typed/src/main/scala/akka/typed/javadsl/BehaviorBuilder.scala:
```diff
@@ -0,0 +1,51 @@
+          case Nil ⇒
+            // emulate scala match error
+            throw new MatchError(msg)
+          case (cls, impl) :: tail ⇒
```
Please sanity check if this doesn't allocate an additional tuple in the deconstructing. Not entirely sure if 2.12 handles this nicely (forgot if this is fixed by default nowadays)

raboof replied:
What would be a good way to check? I looked at the bytecode:
```
  private <M> akka.typed.Behavior<T> receive(akka.typed.javadsl.ActorContext<T>, M, scala.collection.immutable.List<scala.Tuple2<java.lang.Class<? extends M>, scala.Function2<akka.typed.javadsl.ActorContext<T>, M, akka.typed.Behavior<T>>>>);
    Code:
       0: aload_3
       1: astore        5
       3: getstatic     #43                 // Field scala/collection/immutable/Nil$.MODULE$:Lscala/collection/immutable/Nil$;
       6: aload         5
       8: invokevirtual #49                 // Method java/lang/Object.equals:(Ljava/lang/Object;)Z
      11: ifeq          23
      14: new           #51                 // class scala/MatchError
      17: dup
      18: aload_2
      19: invokespecial #55                 // Method scala/MatchError."<init>":(Ljava/lang/Object;)V
      22: athrow
      23: aload         5
      25: instanceof    #57                 // class scala/collection/immutable/$colon$colon
      28: ifeq          119
      31: aload         5
      33: checkcast     #57                 // class scala/collection/immutable/$colon$colon
      36: astore        7
      38: aload         7
      40: invokevirtual #61                 // Method scala/collection/immutable/$colon$colon.head:()Ljava/lang/Object;
      43: checkcast     #63                 // class scala/Tuple2
      46: astore        8
      48: aload         7
      50: invokevirtual #66                 // Method scala/collection/immutable/$colon$colon.tl$1:()Lscala/collection/immutable/List;
      53: astore        9
      55: aload         8
      57: ifnull        119
      60: aload         8
      62: invokevirtual #69                 // Method scala/Tuple2._1:()Ljava/lang/Object;
      65: checkcast     #71                 // class java/lang/Class
      68: astore        10
      70: aload         8
      72: invokevirtual #74                 // Method scala/Tuple2._2:()Ljava/lang/Object;
      75: checkcast     #76                 // class scala/Function2
      78: astore        11
      80: aload         10
      82: aload_2
      83: invokevirtual #80                 // Method java/lang/Object.getClass:()Ljava/lang/Class;
      86: invokevirtual #84                 // Method java/lang/Class.isAssignableFrom:(Ljava/lang/Class;)Z
      89: ifeq          109
      92: aload         11
      94: aload_1
      95: aload_2
      96: invokeinterface #88,  3           // InterfaceMethod scala/Function2.apply:(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
     101: checkcast     #90                 // class akka/typed/Behavior
     104: astore        6
     106: aload         6
     108: areturn
     109: aload_1
     110: aload_2
     111: aload         9
     113: astore_3
     114: astore_2
     115: astore_1
     116: goto          0
     119: new           #51                 // class scala/MatchError
     122: dup
     123: aload         5
     125: invokespecial #55                 // Method scala/MatchError."<init>":(Ljava/lang/Object;)V
     128: athrow
```
Looks like (40-82) no new `Tuple` is allocated (though does show we might want to move `case Nil` down and replace it with `case _`)

ktoso replied:
Yes looks like it does not allocate anything.

ktoso commented on akka-typed-tests/src/test/java/akka/typed/javadsl/BehaviorBuilderTest.java:
```diff
@@ -0,0 +1,35 @@
+                    System.out.println("Terminating along with " + t.ref());
+                    return stopped();
+                })
+                .build();
```
Would be API lovely to not need the `build` call, just using `List` though then won't do as we'll need to keep source code order of the matches (so traversing from the tail), we can work around that though I think

raboof replied:
We could fairly easily make `BehaviorBuilder` implement `ExtensibleBehavior` directly. Calling it a `Builder` might be kind of confusing then though - perhaps `BehaviorChain` or something? Indeed not needing the `.build()` everywhere would be neat.

ktoso replied:
Yeah interesting name suggestion

raboof committed:
remove build(), add methods, predicates

ktoso commented on akka-typed/src/main/scala/akka/typed/javadsl/BehaviorBuilder.scala:
```diff
@@ -5,47 +5,137 @@
 
 object BehaviorBuilder {
-  def create[T](): BehaviorBuilder[T] = BehaviorBuilder[T](Nil, Nil)
+  def apply[T](): BehaviorBuilder[T] = BehaviorBuilder[T](Nil, Nil)
```
either mark as internal, and people should only call `Actor...` or add `create()` (I vote for adding create in case someone finds the builder)

ktoso commented on akka-typed/src/main/scala/akka/typed/javadsl/BehaviorBuilder.scala:
```diff
@@ -5,47 +5,137 @@
+  import scala.language.existentials
+
+  @InternalApi
+  case class Case[BT, MT](`type`: Class[_ <: MT], predicate: Option[MT ⇒ Boolean], handler: (ActorContext[BT], MT) ⇒ Behavior[BT])
```
final + private[akka]?

ktoso commented on akka-typed/src/main/scala/akka/typed/javadsl/BehaviorBuilder.scala:
```diff
@@ -5,47 +5,137 @@
 
 }
 
 object BehaviorBuilder {
```
by making BehaviorBuilder extend it's trait we could gain entry to the builder DSL from any of its methods

ktoso replied:
```
return signalEquals( ...).message

 // or 

return message(...).signal(...)
```

raboof replied:
Hmm right now the generic type of the builder is fixed.

If we'd want to make `object BehaviorBuilder` a common starting point we might be able to start from `BehaviorBuilder[Nothing]` and then potentially widen the generic type at each step... though that might make the types more complicated and compilation error messages when types don't line up less easy to read. Not sure if that'd be worth it.

ktoso replied:
Hm yeah, some experimentation needed - I guess manually replicating the "entry point" methods onto the static context is fine too

ktoso commented on akka-typed-tests/src/test/java/akka/typed/javadsl/BehaviorBuilderTest.java:
```diff
@@ -0,0 +1,59 @@
+
+    @Test
+    public void shouldCompile() {
+        Behavior<Message> b = BehaviorBuilder.<Message>apply()
```
seems like a leak of scaladsl style

ktoso commented on akka-typed-tests/src/test/java/akka/typed/javadsl/BehaviorBuilderTest.java:
```diff
@@ -0,0 +1,59 @@
+              .signal(Terminated.class, (ctx, t) -> {
+                System.out.println("Terminating along with " + t.ref());
+                return stopped();
+              });
```
pretty nice, I wonder if those would be easier to understand if called onMessage / onSignal hm... I guess others will have opinions on this too

ktoso commented on akka-typed/src/main/scala/akka/typed/javadsl/BehaviorBuilder.scala:
```diff
@@ -0,0 +1,141 @@
+package akka.typed.javadsl
+
+import scala.annotation.tailrec
+import akka.japi.pf.FI
```
I think that's the wrong function types to use. Those are from 2014, most of them are not even partial functions but defined in `pf` hm...

I think these should use `import akka.japi.{ Pair, function }` instead

raboof replied:
Seems to be quite some overlap between `akka.japi.*` and `akka.japi.function.*` too - the latter is preferred? let's perhaps add @deprecated notices in another PR?

ktoso replied:
Yeah, we historically made a mess there... some of these did not have variance annotations or throws things... Not sure we can make them deprecated since they may be used in existing non deprecated APIs and we can't add overloads since that will break the target typing of java lambdas... would be good to make a ticket for this - can you open one? About unifying all the functiona types.

ktoso commented on akka-typed/src/main/scala/akka/typed/javadsl/BehaviorBuilder.scala:
```diff
@@ -0,0 +1,141 @@
+   * @tparam M
+   * @return a new behavior with the specified handling appended
+   */
+  def signal[M <: Signal](`type`: Class[M], predicate: FI.TypedPredicate[M], handler: FI.Apply2[ActorContext[T], M, Behavior[T]]): BehaviorBuilder[T] =
```
This should be `akka.japi.function.Predicate` and call the param `test` perhaps?

ktoso commented on akka-typed/src/main/scala/akka/typed/javadsl/BehaviorBuilder.scala:
```diff
@@ -0,0 +1,141 @@
+   * @param `type` type of signals to match
+   * @param predicate a predicate that will be evaluated on the argument if the type matches
+   * @param handler action to apply if the type matches and the predicate returns true
+   * @tparam M
```
docs here

ktoso commented on akka-typed/src/main/scala/akka/typed/javadsl/BehaviorBuilder.scala:
```diff
@@ -0,0 +1,141 @@
+   * @param handler action to apply when the message matches
+   * @return a new behavior with the specified handling appended
+   */
+  def signalEquals(msg: Signal, handler: FI.Apply[ActorContext[T], Behavior[T]]): BehaviorBuilder[T] =
```
`sig: Signal`?

raboof committed:
akka.japi.function, docs, construction

ktoso commented on akka-typed/src/main/scala/akka/typed/javadsl/BehaviorBuilder.scala:
```diff
@@ -0,0 +1,141 @@
+        else receive[M](ctx, msg, tail)
+      case _ ⇒
+        // emulate scala match error
+        throw new MatchError(msg)
```
I may have missed something, why are we doing the match error manually instead allowing scala to throw it for us?

raboof replied:
We're pattern-matching on the list of cases here - but if none of the cases matches, we want to signal that the _message_ didn't match any of the cases. If we'd skip this case we'd get a `MatchError` for the empty list (as we exhausted all cases).

raboof replied:
Though come think of it we probably want to return `unhandled` in this case anyway :)

ktoso commented on akka-typed/src/main/scala/akka/typed/javadsl/BehaviorBuilder.scala:
```diff
@@ -0,0 +1,141 @@
+    }
+
+  /**
+   * Add a new case to the message handling.
```
In final docs of this we should also explain that it matches in order of source code definition I guess?
Like a Scala match. May be obvious, but let's document it as that is the expected behaviour (not that someone asks if it's "most specific class, then the less specific ones etc")

ktoso commented on akka-typed/src/main/scala/akka/typed/javadsl/BehaviorBuilder.scala:
```diff
@@ -0,0 +1,141 @@
+    copy[T](messageHandlers = messageHandlers :+ Case[T, T](msg.getClass, Some(_.equals(msg)), (ctx: ActorContext[T], _: T) ⇒ handler.apply(ctx)))
+
+  /**
+   * Add a new case to the signal handling
```
missing `.` at end - to be consistent with rest of docs

ktoso commented on akka-typed/src/main/scala/akka/typed/javadsl/BehaviorBuilder.scala:
```diff
@@ -0,0 +1,141 @@
+  /**
+   * Add a new case to the signal handling
+   *
+   * @param `type` type of signal to match
```
does this look good in javadoc? (try genjavadoc or just rename to clazz or something similar?)

ktoso commented on akka-typed/src/main/scala/akka/typed/javadsl/BehaviorBuilder.scala:
```diff
@@ -0,0 +1,141 @@
+
+  import scala.language.existentials
+
+  @InternalApi
```
needs comment too, `/** INTERNAL API */`

ktoso commented on akka-typed/src/main/scala/akka/typed/javadsl/BehaviorBuilder.scala:
```diff
@@ -0,0 +1,141 @@
+}
+
+object BehaviorBuilder {
+  def create[T](): BehaviorBuilder[T] = BehaviorBuilder[T](Nil, Nil)
```
ah right, good, apply not needed indeed; good rename

raboof committed:
Javadocs, stand-alone entrypoints

raboof committed:
No case class to pollute the javadocs less

patriknw commented on akka-typed-tests/src/test/java/akka/typed/javadsl/BehaviorChainTest.java:
```diff
@@ -0,0 +1,45 @@
+
+    @Test
+    public void shouldCompile() {
+      Behavior<Message> b = Actor.<Message>immutable()
```
`Actor.<Message>immutable` is a syntax that I think is rather unfamiliar for most Java devs. Without that explicit type the type inference doesn't work.

Here is an alternative:
```
def immutable2[T](b: JFunction[BehaviorChain[T], BehaviorChain[T]]): Behavior[T] = ???
```
that is used like:
```
      Behavior<Message> b = Actor.immutable2(b2 ->
        b2.message(One.class, (ctx, o) -> {
          o.foo();
          return same();
        })
        .message(One.class, o -> o.foo().startsWith("a"), (ctx, o) -> same())
      );
```

another advantage of that is that the builder passed as parameter and there is no question about how to create it, and it could be a real builder that we call `.build` on afterwards

patriknw replied:
I think there can be advantages of having a real builder, such as performance optimizations. E.g. if the builder only contains exact message class matches it can can use a HashMap instead of traversing the matches.

Another thing we must consider is the order of the matches. The above example will match on `startsWith("a")` first even though we read it from top-down. I think we must use the same order as in the AbstractActor `PFBuilder`. With a real builder we have more freedom for efficiently implementing such things also.

ktoso replied:
Nice improvement, good idea on passing in the chain inside there!

johanandren replied:
In streams we use a class parameter as an alternative to just a type parameter, which would be: `immutable(Protocol.class).match(etc)`

I'm curious why Patriks function-with-parameter-with-type-parameter would be inferred by the Java compiler if the one in this PR isn't.

patriknw replied:
We will also need similar type of matching in MutableBehavior, but it will have to be a different implementation since the ctx parameter is not used there.

raboof replied:
While I think accepting a `Builder => Builder` will also make most Java devs do a double-take, I do like how this makes it a 'normal' builder again without adding `.build()` all over the place. I'll give it a go!

johanandren replied:
Given that we already have `.build()` in the Java APIs and I haven't heard anyone complain about that I'm not sure i agree that is a problem. I think the higher order builder function part might be a bigger problem.

raboof replied:
Tried it out in https://github.com/akka/akka/pull/22834/files - I think the non-higher-order, while not as short, does seem more straightforward.

patriknw replied:
I agree

patriknw commented on akka-typed/src/main/scala/akka/typed/javadsl/Actor.scala:
```diff
@@ -168,6 +168,20 @@ object Actor {
   }
 
   /**
+   * Constructs an actor behavior builder that can build a behaviour that can react to both
```
let's stick to american spelling, behavior

patriknw commented on akka-typed/src/main/scala/akka/typed/javadsl/BehaviorChain.scala:
```diff
@@ -0,0 +1,243 @@
+   * @return a new behavior with the specified handling appended
+   */
+  def message[M <: T](`type`: Class[M], handler: Function2[ActorContext[T], M, Behavior[T]]): BehaviorChain[T] =
+    new BehaviorChain[T](messageHandlers :+ Case[T, T](`type`, None, (i1: ActorContext[T], msg: T) ⇒ handler.apply(i1, msg.asInstanceOf[M])), signalHandlers)
```
use `Vector` if appending, `List` is for prepending, otherwise we could use real builder and reverse

patriknw commented on akka-typed/src/main/scala/akka/typed/javadsl/BehaviorChain.scala:
```diff
@@ -0,0 +1,243 @@
+   * @return a new behavior with the specified handling appended
+   */
+  def signal[M <: Signal](`type`: Class[M], test: Predicate[M], handler: Function2[ActorContext[T], M, Behavior[T]]): BehaviorChain[T] =
+    new BehaviorChain[T](messageHandlers, signalHandlers :+ Case[T, Signal](`type`, Some((t: Signal) ⇒ test.test(t.asInstanceOf[M])), (ctx: ActorContext[T], signal: Signal) ⇒ handler.apply(ctx, signal.asInstanceOf[M])))
```
very long line, break lines when exceeding 120 chars

patriknw commented on akka-typed/src/main/scala/akka/typed/javadsl/BehaviorChain.scala:
```diff
@@ -0,0 +1,243 @@
+   * @tparam M type of message to match
+   * @return a new behavior with the specified handling appended
+   */
+  def message[T, M <: T](`type`: Class[M], test: Predicate[M], handler: Function2[ActorContext[T], M, Behavior[T]]): BehaviorChain[T] =
```
I'm not sure these "static" methods are worth it, let's see when we have decided how the `BehaviorChain` (builder) can be created

johanandren commented on akka-typed/src/main/scala/akka/typed/javadsl/BehaviorChain.scala:
```diff
@@ -0,0 +1,243 @@
+ *
+ * @tparam T the common superclass of all supported messages.
+ */
+class BehaviorChain[T] private (
```
What was the rationale behind not reusing the existing `akka.japi.JavaPartialFunction` which is on top of the Scala partial function impl?

patriknw replied:
I don't know if it was the rationale but I think we can do better than the japi PF. It's not difficult stuff. e.g. partial function chaining is inefficient.

Another reason might be the ctx param, but of course we could do only message matching.

patriknw replied:
one thing I would like to support is efficient match on exact message class by lookup in HashMap, that is what you need in 95% of all cases

johanandren replied:
Alright, thanks for clarifying

raboof committed:
Make BehaviorChain a proper BehaviorBuilder again

Separating these steps will keep us more flexible to introduce optimizations
later, at the expense of not being able to use the BehaviorBuilder as a
Behavior directly

raboof committed:
Less awkward generic when constructing the builder from Java

patriknw commented on akka-typed-tests/src/test/java/akka/typed/javadsl/BehaviorBuilderTest.java:
```diff
@@ -0,0 +1,46 @@
+
+    @Test
+    public void shouldCompile() {
+      Behavior<Message> b = Actor.immutable(Message.class)
```
looking at this and the [higher-order syntax](https://github.com/akka/akka/pull/22834) I think this is probably easier to understand

ktoso replied:
+1, this has the most "java feel" to it.

johanandren replied:
I like this one best as well.

patriknw commented on akka-typed/src/main/scala/akka/typed/javadsl/BehaviorBuilder.scala:
```diff
@@ -0,0 +1,245 @@
+  private val signalHandlers:  List[Case[T, Signal]]
+) {
+
+  def build(): Behavior[T] = new ExtensibleBehavior[T] {
```
I think this is good, gives us some flexibility

raboof committed:
Shorten/DRYer implementation for adding message/signal handlers

raboof committed:
Introduce 'mutable' javadsl builder API

ktoso commented on akka-typed-tests/src/test/java/akka/typed/javadsl/BehaviorBuilderTest.java:
```diff
@@ -0,0 +1,99 @@
+    @Test
+    public void shouldCompile() {
+      Behavior<Message> b = Actor.immutable(Message.class)
+              .message(One.class, (ctx, o) -> {
```
those will be onMessage though, right?

raboof replied:
works for me, I don't have a strong preference either way.

patriknw replied:
seems we are using `onMessage` and `onSignal` in other places so makes sense here also

raboof commented:
@patriknw I experimented with extending `MutableBehavior` and making this look more like `AbstractActor` but wasn't too happy with the result yet (https://github.com/akka/akka/pull/22836) - thoughts?

raboof committed:
Use onMessage/onSignal in the builder as well

raboof committed:
Return unhandled if no case matches

raboof committed:
Using builder but extend MutableBehavior

* Use 'receive' terminology
* Use 'onMessage'/'onSignal' for receive builder, too
* 'unhandled' when no case matches
* Receive interface, promote 'this' for mutable behaviors
* Move ReceiveBuilder tests to its own file
* Remove need for StateHolder
* Test for and fix initialization issue
* Avoid lazy val during mutable actor initialization

patriknw reviewed (approved):
LGTM, great work
